diff --git a/video/video_receive_stream.cc b/video/video_receive_stream.cc
index 9e94e5e94e..8bd0f8a9c7 100644
--- a/video/video_receive_stream.cc
+++ b/video/video_receive_stream.cc
@@ -337,9 +337,11 @@ void VideoReceiveStream::Start() {
   }
 
   for (const Decoder& decoder : config_.decoders) {
-    std::unique_ptr<VideoDecoder> video_decoder =
-        config_.decoder_factory->LegacyCreateVideoDecoder(decoder.video_format,
-                                                          config_.stream_id);
+    std::unique_ptr<VideoDecoder> video_decoder = NULL;
+    if (!getenv("USE_NULL_VIDEO_DECODER")) {
+        video_decoder = decoder.decoder_factory->LegacyCreateVideoDecoder(decoder.video_format,
+                                                                          config_.stream_id);
+    }
     // If we still have no valid decoder, we have to create a "Null" decoder
     // that ignores all calls. The reason we can get into this state is that the
     // old decoder factory interface doesn't have a way to query supported
diff --git a/video/video_receive_stream2.cc b/video/video_receive_stream2.cc
index ff31c6613a..f8f6d80cde 100644
--- a/video/video_receive_stream2.cc
+++ b/video/video_receive_stream2.cc
@@ -414,8 +414,10 @@ void VideoReceiveStream2::Stop() {
 
 void VideoReceiveStream2::CreateAndRegisterExternalDecoder(
     const Decoder& decoder) {
-  std::unique_ptr<VideoDecoder> video_decoder =
-      config_.decoder_factory->CreateVideoDecoder(decoder.video_format);
+  std::unique_ptr<VideoDecoder> video_decoder = NULL;
+  if (!getenv("USE_NULL_VIDEO_DECODER")) {
+    video_decoder = config_.decoder_factory->CreateVideoDecoder(decoder.video_format);
+  }
   // If we still have no valid decoder, we have to create a "Null" decoder
   // that ignores all calls. The reason we can get into this state is that the
   // old decoder factory interface doesn't have a way to query supported
